src/model/history.js

```diff
@@ -36,109 +35,176 @@ export default class History {
-		if ( delta && this._deltas[ this._deltas.length - 1 ] !== delta ) {
-			const index = this._deltas.length;
+	*getDeltas( from = 0, to = Number.POSITIVE_INFINITY ) {
+		// No deltas added, nothing to yield.
+		if ( this._deltas.length === 0 ) {
+			return;
+		}
 
-			this._deltas[ index ] = delta;
-			this._historyPoints.set( delta.baseVersion, index );
+		// Will throw if base version is incorrect.
+		let fromIndex = this._getIndex( from );
+
+		// Base version is too low or too high and is not found in history.
+		if ( fromIndex == -1 ) {
+			return;
+		}
+
+		// We have correct `fromIndex` so let's iterate starting from it.
+		while ( fromIndex < this._deltas.length ) {
+			const delta = this._deltas[ fromIndex++ ];
+
+			if ( delta.baseVersion >= to ) {
+				break;
+			}
+
+			yield delta;
 		}
 	}
 
 	/**
-	 * Transforms out-dated delta by all deltas that were added to the history since the given delta's base version. In other
-	 * words, it makes the delta up-to-date with the history. The transformed delta(s) is (are) ready to be applied
-	 * to the {@link engine.model.Document document}.
+	 * Returns one or more deltas from history that bases on given `baseVersion`. Most often it will be just
+	 * one delta, but if that delta got updated by multiple deltas, all of those updated deltas will be returned.
 	 *
-	 * @param {engine.model.delta.Delta} delta Delta to update.
-	 * @returns {Array.<engine.model.delta.Delta>} Result of transformation which is an array containing one or more deltas.
+	 * @see engine.model.History#updateDelta
+	 * @param {Number} baseVersion Base version of the delta to retrieve.
+	 * @returns {Array.<engine.model.delta.Delta>|null} Delta with given base version or null if no such delta is in history.
 	 */
-	getTransformedDelta( delta ) {
-		if ( delta.baseVersion === this._nextHistoryPoint ) {
-			return [ delta ];
+	getDelta( baseVersion ) {
+		let index = this._historyPoints.get( baseVersion );
+
+		if ( index === undefined ) {
+			return null;
 		}
 
-		let transformed = [ delta ];
+		const deltas = [];
 
-		for ( let historyDelta of this.getDeltas( delta.baseVersion ) ) {
-			let allResults = [];
+		for ( index; index < this._deltas.length; index++ ) {
+			const delta = this._deltas[ index ];
 
-			for ( let deltaToTransform of transformed ) {
-				const transformedDelta = History._transform( deltaToTransform, historyDelta );
-				allResults = allResults.concat( transformedDelta );
+			if ( delta.baseVersion != baseVersion ) {
+				break;
 			}
 
-			transformed = allResults;
+			deltas.push( delta );
 		}
 
-		// Fix base versions.
-		let baseVersion = transformed[ 0 ].operations[ 0 ].baseVersion;
-
-		for ( let i = 0; i < transformed.length; i++ ) {
-			transformed[ i ].baseVersion = baseVersion;
-			baseVersion += transformed[ i ].operations.length;
-		}
+		return deltas.length === 0 ? null : deltas;
+	}
```

Comment:
It should be fine. AFAICS there are three things that should be considered:
1) Transforming incoming deltas.
2) Sending changes to other clients.
3) Repeating history when new client joins.

1) There are several cases here to consider. As long as you have correct state in compressed history and correctly "pre-transform" incoming operation it will be fine. I will write proper explanation at the bottom of this post. Anyway, the operational/delta transformations got a bit complicated and their base versions are fixed on multiple levels. I actually think it could be a good idea to get rid of fixing (incrementing) base version in OT algorithms. A feature that transforms deltas should be responsible for setting proper base version. 

2) Deltas are created and applied by `Batch` API. Nothing changed there. Each created delta will have a new, next number. OT plugin will listen for them and propagate them to server (BTW. I thought about it yesterday and I don't think `change` event is enough for this -- we might consider introducing `delta` event and/or change `engine.model.Document#applyOperation` to `#applyDelta`). This means that compressed history having multiple deltas with same base version are enclosed for each client.

3) When new client joins, it's editor contents have to be initialized. But you don't have to repeat history to do so. All you need is to make a "snapshot" of current model state and pass it as one big initial `InsertDelta`. Or you just send the "snapshot" itself and OT plugin on client side will convert it to initial `InsertDelta` (or just use `setData` method). All you need to do afterwards is to set `engine.model.Document#baseVersion` to correct value.


Here are some cases of what might happen when you get incoming delta while your history got compressed. Let's assume given history stack. `----` means that there was delta there but it got removed.
```
   A
   B
---C1---
---C2---
   Da without C
   Db without C
   E without C
---C1'---
---C2'---
   F
```
What happened here is that batch `C` got reverted and removed. Delta `D` and `E` are updated, but `D` got updated into two deltas (keep in mind this is a very rare scenario and I am not sure it could even happen :)). Delta `F` was added after undo happened.

Let's assume there is incoming, external delta `X`. Remember that the remote client will not have any updated/removed deltas, just the originals. Let's crack down following scenarios. It could base on:
1) delta `A`, `B`, or any delta before `C1`.
2) delta `C1`
3) delta `C2`
4) delta `D`, `E`, or any delta between `C2` and `C1'`
5) delta `C2'`
6) delta `F`.

Keep in mind, that the delta can't base on `Da` or `Db` because these are updated deltas and we don't send them to the server/other clients.

*In fact, as I think of it now, it could also base on `C1'`, which I haven't anticipated, because `C1'` and `C2'` are applied one after another, but they don't have to happen right after another on other client machine. Maybe we should/will have to introduce some control mechanisms, that one delta can be send/applied only together with other deltas?*

Here are solutions:
1) There are no changed deltas above, so both client's histories were same when delta `X` got created. We can simply transform delta `X` by history. It's the same case as when undoing deltas.

2) Delta `X` bases on delta `C1` that got reversed and removed from history. Deltas below remember the state "without delta `C1`". We have to convert delta `X` to the same state. To do so, we have to convert it by reversed `C1` - this way it will "forget" about `C1`. After delta `X` is transformed to `X without C` it can be normally transformed by compressed history state. Keep in mind that converting by reversed deltas is not an efficiency cost -- because we will not transform `X` later by `C1'` etc. And we would have to do so if we keep original history. Also notice that it's what we do normally while undoing deltas.

3) and 4) are actually the same scenarios as 2). We have to transform by all previously reversed deltas to get `X` to the proper state.

5) and 6) delta `X` is already in a state after `C` happened and after `C` got reversed, so there is no need to do any special things.

*If `X` bases on `C1'` it would need to be converted by `C2'` and all following deltas. So this would be another case, but similar to 2), 3) and 4). Anyway, for all those cases to work, we need to keep a registry of original/removed deltas, because we will be in need of transforming by them.*